onnx/checker.cc

```diff
@@ -0,0 +1,247 @@
+  enforce_non_empty_field(attr, name);
+
+  if (ir_version >= 0x00000002) {
+    enforce_has_field(attr, type);
+  }
+
+  int used_fields = 0;
+
+#define check_type(expected_type)                        \
+  if (attr.has_type() && attr.type() != expected_type) { \
+    fail("type field and data field mismatch.");         \
+  }
+
+#define check_singular_field(field, type) \
+  if (attr.has_##field()) {               \
+    ++used_fields;                        \
+    check_type(type);                     \
+  }
+
+#define check_repeated_field(field, type) \
+  if (attr.field##_size() > 0) {          \
+    ++used_fields;                        \
+    check_type(type);                     \
+  }
+
+  check_singular_field(f, AttributeProto::FLOAT);
+  check_singular_field(i, AttributeProto::INT);
+  check_singular_field(s, AttributeProto::STRING);
+  check_singular_field(t, AttributeProto::TENSOR);
+  check_singular_field(g, AttributeProto::GRAPH);
+  check_repeated_field(floats, AttributeProto::FLOATS);
+  check_repeated_field(ints, AttributeProto::INTS);
+  check_repeated_field(strings, AttributeProto::STRINGS);
+  check_repeated_field(tensors, AttributeProto::TENSORS);
+  check_repeated_field(graphs, AttributeProto::GRAPHS);
+
+#undef check_type
+#undef check_singular_field
+#undef check_repeated_field
+
+  if (used_fields != 1) {
+    fail("Attribute should contain one and only one value field.");
+  }
+
+  for (const auto& tensor : attr.tensors()) {
+    check_tensor(tensor, ir_version);
+  }
+  for (const auto& graph : attr.graphs()) {
+    check_graph(graph, ir_version);
+  }
+}
+
+void check_node(const NodeProto& node, int ir_version) {
+  enforce_non_empty_field(node, op_type)
+
+      if (node.input().empty() && node.output().empty()) {
+    fail("NodeProto has zero input and zero output.");
+  }
+
+  for (const auto& attr : node.attribute()) {
+    check_attribute(attr, ir_version);
+  }
+
+  const auto* schema = OpSchemaRegistry::Schema(node.op_type());
+  if (!schema) {
+    fail("No Schema registered for " + node.op_type());
+  }
+  if (!schema->Verify(node)) {
+    fail("NodeProto of type " + node.op_type() + " did not pass schema check.");
+  }
+}
+
+void check_graph(const GraphProto& graph, int ir_version) {
+  enforce_non_empty_field(graph, name);
+
+  for (const auto& value_info : graph.input()) {
+    check_value_info(value_info, ir_version);
+  }
+  for (const auto& value_info : graph.output()) {
+    check_value_info(value_info, ir_version);
+  }
+  for (const auto& node : graph.node()) {
+    check_node(node, ir_version);
+  }
+
+  std::unordered_set<std::string> input_names{};
+  for (const auto& value_info : graph.input()) {
+    input_names.insert(value_info.name());
+  }
+  for (const auto& init : graph.initializer()) {
+    if (!input_names.count(init.name())) {
+      fail(init.name() + " in initializer but not in graph input");
+    }
+    check_tensor(init, ir_version);
+  }
+}
```

Comment:
we should add here proper checking of the graph, i.e. that there are no unconnected inputs, duplicated names, etc.

Reply:
will fix it